File: androidshared/src/main/java/org/odk/collect/androidshared/network/ConnectivityProvider.java
package org.odk.collect.androidshared.network;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

public class ConnectivityProvider implements NetworkStateProvider {

    private final Context context;

    public ConnectivityProvider(Context context) {
        this.context = context;
    }

    public boolean isDeviceOnline() {
        NetworkInfo networkInfo = getNetworkInfo();
        return networkInfo != null && networkInfo.isConnected();
    }

    public NetworkInfo getNetworkInfo() {
        return getConnectivityManager().getActiveNetworkInfo();
    }

    private ConnectivityManager getConnectivityManager() {
        return (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
    }
}

File: androidshared/src/main/java/org/odk/collect/androidshared/network/ConnectivityProvider.kt
package org.odk.collect.androidshared.network

import android.content.Context
import android.net.ConnectivityManager
import org.odk.collect.async.Scheduler

class ConnectivityProvider(private val context: Context) : NetworkStateProvider {
    override val isDeviceOnline: Boolean
        get() {
            val activeNetworkInfo = connectivityManager.activeNetworkInfo
            return activeNetworkInfo != null && activeNetworkInfo.isConnected
        }

    override val currentNetwork: Scheduler.NetworkType?
        get() {
            return when (connectivityManager.activeNetworkInfo?.type) {
                ConnectivityManager.TYPE_WIFI -> Scheduler.NetworkType.WIFI
                ConnectivityManager.TYPE_MOBILE -> Scheduler.NetworkType.CELLULAR
                else -> null
            }
        }

    private val connectivityManager: ConnectivityManager
        get() = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
}
